### PR TITLE
Replace Moodycamel's queue with Staging's rwqueue

### PR DESCRIPTION
**Pull request by kcgen**
The Moodycamel queue was replaced with Staging's rwqueue in a line-for-line replacement.
 
# Benchmarks

The same callback-latency procedure was used as described here - https://github.com/dosbox-staging/dosbox-staging/pull/887

All times are in microseconds. Lower is better.

| Branch | Vector handling | median | avg | stdev | Data |
|  --- | --- | ---: | ---: | ---: | --- |
| moodycamel | copy | 2.12  |   3.89  |  26.11 | Previously reported [here](https://github.com/dosbox-staging/dosbox-staging/pull/906#issuecomment-782879076) |
| moodycamel | std::move | 1.87   |  3.57  |  10.82 | Previously reported [here](https://github.com/dosbox-staging/dosbox-staging/pull/906#issuecomment-782879076) |
| staging rwqueue | std::move | 1.79  |   3.17  |   7.88 | [pr-use-rwqueue-1.tar.gz](https://github.com/dosbox-staging/dosbox-staging/files/6077279/pr-use-rwqueue-1.tar.gz) |

## Results

I can't explain why our queue produces even lower latency real-world results given Moodycamel's queue objectively benchmarks at roughly twice the bulk-throughput rate. 

It could be that our queue is smaller and simpler, resulting in a better overall real-world instruction cache hit-rate when running amidst large amounts of other code (versus being given the entire CPU's L1/L2 cache to run in a tight loop like the queue benchmark performs).

I did three runs just to be sure  (all in the tar gz). Whatever the reason, in the end we're not any slower.  I'll take it!

### CODE REVIEW

File: src/midi/midi_fluidsynth.cpp [hunk 1]
```diff
@@ -175,7 +175,9 @@ static std::string find_sf_file(const std::string &name)
 
 MidiHandlerFluidsynth::MidiHandlerFluidsynth()
         : soft_limiter("FSYNTH", prescale_level, FRAMES_PER_BUFFER)
-{}
+{
+	keep_rendering = false;
```
Comment (kcgen) [inline src/midi/midi_fluidsynth.cpp:179]: Atomic types normal don't allow an initial value, however I believe the atomic book is unique in this regard, so I might be able to move this into the constructor's initialization block.  Will try.

Comment (dreamer) [reply src/midi/midi_fluidsynth.cpp:179]: Just note: atomic values are not necessarily faster than values hidden behind mutexes - they are just atomic.

**Comment by dreamer** [comment]
> I can't explain why our queue produces even lower latency real-world results given Moodycamel's queue objectively benchmarks at roughly twice the bulk-throughput rate.
> 
> It could be that our queue is smaller and simpler, resulting in a better overall real-world instruction cache hit-rate when running amidst large amounts of other code (versus being given the entire CPU's L1/L2 cache to run in a tight loop like the queue benchmark performs).

I rather suspect that stupid userspace-level spinlock implementation was the problem. Normal mutexes are just as fast or faster when used correctly, so I guess that's what we might be seeing here :)

Our rwqueue has very few elements as compared to normal usecase for moodycamel's queue (initially I thought we are going to use it with thousands of samples and not few buffers), so additional overhead of that implementation might be the cause. Or simply that queue implementation was designed to win in few synthetic benchmarks, perhaps only on Windows.

I guess we should ask for re-test on that Pine64 ARM system that was having problems earlier…

File: src/midi/midi_fluidsynth.cpp (hunk 1 above)
Comment (kcgen) [reply src/midi/midi_fluidsynth.cpp]: > Just note: atomic values are not necessarily faster than values hidden behind mutexes - they are just atomic.

Indeed; they are sure nicer to work with though (in the form of a single variable, with mutual-access guaranteed behind the scene), versus carrying a mutex around and needing several lines of code to read or write values.

Comment (kcgen) [reply src/midi/midi_fluidsynth.cpp]: Also good news - the atomic_bool can be initialized in the constructor.  Update inbound.

**Comment by kcgen** [comment]
> I guess we should ask for re-test on that Pine64 ARM system that was having problems earlier…

@SupervisedThinking -- can you give `kc/use-rwqueue-1` a go on your RockPro64? We've updated the buffer and thread pacing using a drop-in replacement for the prior queue we were using.  Thank you!

**Comment by SupervisedThinking** [comment]
- RK3399 / ROCKPro64

Fluidsynth
```
dosbox-staging version 8.95.002-6535-g4515c3a2ba
---
CONFIG: Loading primary file /storage/roms/dos/testARM/Dark_Forces_FS_SF3_MuseScore_General.conf
CONFIG: Unknown option glshader
CONFIG: Unknown option synth_threads
MAIN: Draw resolution: 640x480, pixel aspect ratio: 1.00
SDL: Using driver "opengles2" for texture renderer
SDL: Mouse will move seamlessly without being captured.
MEMORY: Base address: 0xf006e008
MEMORY: Using 4096 DOS memory pages (16 MiB)
MIXER: Negotiated 2-channel 44100-Hz audio in 1024-frame blocks
MIXER: FSYNTH channel operating at 44100 Hz without resampling
MIDI: Using SoundFont '/storage/.config/soundfonts/MuseScore_General.sf3'
MIDI: Opened device: fluidsynth
MIXER: FM channel operating at 44100 Hz without resampling
MIXER: SB channel operating at 22050 Hz and upsampling to the output rate
MIXER: GUS channel operating at 1 Hz and upsampling to the output rate
GUS: @SET ULTRASND=240,3,3,5,5
MIXER: SPKR channel operating at 18939 Hz and upsampling to the output rate
MIXER: DISNEY channel operating at 10000 Hz and upsampling to the output rate
MAPPER: Found no joysticks
MAPPER: Loaded default key bindings
MOUNT: Path '/storage/roms/dos' found
SDL: captured the mouse
MAIN: Draw resolution: 640x400, pixel aspect ratio: 1.20
SDL: Using driver "opengles2" for texture renderer
DOSBox has switched to max cycles, because of the setting: cycles=auto.
If the game runs too fast, try a fixed cycles amount in DOSBox's options.
SB16: DSP was reset
SB16: Speaker-output has been toggled on
MAIN: Draw resolution: 320x200, double-width, double-height, pixel aspect ratio: 1.20
SDL: Using driver "opengles2" for texture renderer
SB16: DSP was reset
MAIN: Draw resolution: 640x400, pixel aspect ratio: 1.20
SDL: Using driver "opengles2" for texture renderer
FSYNTH: Peak amplitude reached 91% of max
MAIN: Draw resolution: 640x400, pixel aspect ratio: 1.20
SDL: Using driver "opengles2" for texture renderer
SDL: released the mouse
```

**Comment by kcgen** [comment]
Thanks for testing on your devices @SupervisedThinking !

**Review by dreamer — APPROVED** [review]
I did some testing on my side and FluidSynth works well, even after I limited my CPU to ~800MHz per core :)

Interestingly, this PR uses <atomic>, but does not require linking -latomic…

**Comment by dreamer** [comment]
@kcgen Thanks! :)

**Comment by kcgen** [comment]
> I did some testing on my side and FluidSynth works well, even after I limited my CPU to ~800MHz per core :)

:+1:  

...

> Interestingly, this PR uses , but does not require linking -latomic…

Indeed - I'm not sure why this one links but the other doesn't!

**Pull request merged**